include/kamping/distributed_graph_communicator.hpp

```diff
@@ -0,0 +1,353 @@
+    /// integers which can be decomposed by a structured binding via `auto [rank, weight] = weighted_edge;`.
+    ///
+    /// @tparam InEdgeRange Range type of in-going edges.
+    /// @tparam InEdgeRange Range type of out-going edges.
+    /// @param in_edges Range of in-going edges.
+    /// @param out_edges Range of out-going edges.
+    ///
+    template <typename InEdgeRange, typename OutEdgeRange>
+    CommunicationGraph(InEdgeRange const& in_edges, OutEdgeRange const& out_edges) {
+        constexpr bool are_in_edges_weighted  = internal::are_edges_weighted<InEdgeRange>();
+        constexpr bool are_out_edges_weighted = internal::are_edges_weighted<OutEdgeRange>();
+        static_assert(
+            are_in_edges_weighted == are_out_edges_weighted,
+            "Weight status of in and out edges is different!"
+        );
+        if constexpr (are_in_edges_weighted) {
+            auto get_rank = [](auto const& edge) {
+                auto const& [rank, _] = edge;
+                return static_cast<int>(rank);
+            };
+            auto get_weight = [](auto const& edge) {
+                auto const& [_, weight] = edge;
+                return static_cast<int>(weight);
+            };
+            transform_elems_to_integers(in_edges, _in_ranks, get_rank);
+            transform_elems_to_integers(out_edges, _out_ranks, get_rank);
+            _in_weights  = DefaultContainer<int>{};
+            _out_weights = DefaultContainer<int>{};
+            transform_elems_to_integers(in_edges, _in_weights.value(), get_weight);
+            transform_elems_to_integers(out_edges, _out_weights.value(), get_weight);
+        } else {
+            auto get_rank = [](auto const& edge) {
+                return static_cast<int>(edge);
+            };
+            transform_elems_to_integers(in_edges, _in_ranks, get_rank);
+            transform_elems_to_integers(out_edges, _out_ranks, get_rank);
+        }
+    }
+
+    /// @brief Constructs a communication graph based on a range of unweighted in-going and out-going edges.
+    CommunicationGraph(DefaultContainer<int>&& in_ranks, DefaultContainer<int>&& out_ranks)
+        : _in_ranks{std::move(in_ranks)},
+          _out_ranks{std::move(out_ranks)} {}
+
+    /// @brief Constructs a communication graph based on a range of weighted in-going and out-going edges.
+    CommunicationGraph(
+        DefaultContainer<int>&& in_ranks,
+        DefaultContainer<int>&& out_ranks,
+        DefaultContainer<int>&& in_weights,
+        DefaultContainer<int>&& out_weights
+    )
+        : _in_ranks{std::move(in_ranks)},
+          _out_ranks{std::move(out_ranks)},
+          _in_weights{std::move(in_weights)},
+          _out_weights{std::move(out_weights)} {}
+
+    /// @brief Constructs a communication graph based on a range of potentially weighted symmetric edges.
+    template <typename EdgeRange>
```

Comment:
wording: Construct a communication graph where in and out neighbors are the same, i.e. symmetric.